src/thunderbots/software/multithreading/thread_safe_buffer.h

```diff
@@ -0,0 +1,157 @@
+#pragma once
+
+#include <boost/circular_buffer.hpp>
+#include <condition_variable>
+#include <cstddef>
+#include <deque>
+#include <mutex>
+#include <optional>
+
+#include "util/time/duration.h"
+
+/**
+ * This class represents a buffer of objects
+ *
+ * It's public API is fully thread-safe (meaning that it can be accessed by
+ * multiple threads at the same time without issue)
+ *
+ * @tparam T The type of whatever is being buffered
+ */
+template <typename T>
+class ThreadSafeBuffer
+{
+   public:
+    // Force the user to specify a size
+    explicit ThreadSafeBuffer() = delete;
+
+    explicit ThreadSafeBuffer(std::size_t buffer_size);
+
+    // Copying this class is not permitted
+    ThreadSafeBuffer(const ThreadSafeBuffer&) = delete;
+
+    /**
+     * Removes the least recently value added to the buffer and returns it
+     *
+     * If the buffer is empty, this function will *block* until:
+     * - a value becomes available
+     * - the given amount of time is exceeded
+     * - the destructor of this class is called
+     *
+     * @param max_wait_time The maximum duration to wait for a new value before
+     *                      returning
+     *
+     * @return The most recently value added to the buffer, or std::nullopt if none is
+     *         available
+     */
+    std::optional<T> pullLeastRecentlyAddedValue(
+        Duration max_wait_time = Duration::fromSeconds(0));
+
+    /**
+     * Removes the most recently value added to the buffer and returns it
```

Comment:
Replied above, this should be an easy fix once the above comment is figured out.

Reply:
After discussion above, added a quick example to the javadoc: `ex. if A,B,C were added to the buffer (in that order), this would return C` and renamed to `popMostRecentlyAddedValue`